lib/elixir/test/elixir/module/types/pattern_test.exs

```diff
@@ -225,6 +225,11 @@ defmodule Module.Types.PatternTest do
     assert {:ok, :dynamic, context} = quoted_guard([x], elem(x, 0))
     assert Types.lift_type({:var, 0}, context) == :tuple
 
+    assert {:ok, {:atom, true}, context} = quoted_guard([], true)
+    assert {:ok, {:atom, false}, context} = quoted_guard([], false)
+    assert {:ok, {:atom, :fail}, context} = quoted_guard([], :fail)
+    assert {:ok, :boolean, context} = quoted_guard([], is_atom(true or :fail))
```

Comment:
Do we validate though that the guard itself should return `true | false | :fail`? For example, `true or :whatever` as a guard should raise an error (i.e. we will special case only `:fail`).

Reply:
We are doing that in the body checking branch. Can't do it yet because it requires changes to the context and stack maps.

Reply:
Ok!